crates/contracts/solidity/Solver.sol

```diff
@@ -46,19 +50,35 @@ contract Solver {
         address buyToken,
         address nativeToken,
         address payable receiver,
-        bytes calldata settlementCall
+        bytes calldata settlementCall,
+        bool mockPreconditions
     ) external returns (
         uint256 gasUsed,
         uint256[] memory queriedBalances
     ) {
         require(msg.sender == address(this), "only simulation logic is allowed to call 'swap' function");
+
         // Prepare the trade in the context of the trader so we are allowed
         // to set approvals and things like that.
-        Trader(trader).prepareSwap(settlementContract, sellToken, sellAmount, nativeToken, receiver);
+        if (mockPreconditions) {
+            Trader(trader)
+                .prepareSwap(
+                    settlementContract,
+                    sellToken,
+                    sellAmount,
+                    nativeToken
+                );
+        }
+
+        // Warm the storage for sending ETH to smart contract addresses.
+        // We allow this call to revert becaues it was either unnecessary in the first place
+        // or failing to send `ETH` to the `receiver` will cause a revert in the settlement
+        // contract.
+        receiver.call{value: 0}("");
```

Comment:
Was it moved here just as a part of refactoring, or is it essential for the changes to work correctly?

Reply:
Having to warm the storage for sending ETH to a smart contract wallet is a very annoying implementation detail the user shouldn't have to worry about. This part should always happen and should not lead to reverts for any reasonable smart contract receiver.
And since the `prepareSwap()` step is now optional depending on the existence of pre-interactions I moved it out of there. Also the context of this doesn't matter (i.e. whether the trader or the solver sends it).